### PR TITLE
Feature/10 - UserSettings

### PR DESCRIPTION
**Pull request by valerachuk**
image
Can't override parent's !important padding

### CODE REVIEW

File: front/package.json [hunk 1]
```diff
@@ -12,9 +12,9 @@
     "core-js": "^3.6.4",
     "vue": "^2.6.11",
     "vue-i18n": "^8.17.4",
-    "vue-router": "^3.1.6",
-    "vuetify": "^2.2.11",
-    "vuex": "^3.1.3"
+    "vue-router": "^3.2.0",
+    "vuetify": "^2.2.28",
+    "vuex": "^3.4.0"
```
Comment (horba) [inline front/package.json]: ну вообще как-то нехорошо вышло

File: front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html [hunk 2]
```diff
@@ -0,0 +1,3 @@
+<v-btn class="mmt-continue-button" rounded depressed large v-bind="attrs" v-on="$listeners">
+  <slot></slot>
+</v-btn>
```
Comment (horba) [inline front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html]: не думаю, что стоит выносить кнопку в отдельный компонент

File: front/src/config/config.dev.json [hunk 3]
```diff
@@ -1,3 +1,4 @@
 {
-  "serverPath": "https://localhost:5001/"
+  "serverPath": "https://localhost:5001/",
+  "emailRegex": "^\\w+([\\.-]?\\w+)*@\\w+([\\.-]?\\w+)*(\\.\\w{2,3})+$"
```
Comment (horba) [inline front/src/config/config.dev.json]: why here? actually this file is not for those purposes

File: front/src/locales/translations/ua.json [hunk 4]
```diff
@@ -39,5 +39,53 @@
...
+    },
+    "language": {
+      "label": "Мова",
+      "options": {
+        "English": "Англійська",
+        "Russian": "Російська",
+        "Ukrainian": "Українська",
+        "German": "Німецька"
+      }
+    },
```
Comment (horba) [inline front/src/locales/translations/ua.json]: we have had these translations already

File: front/src/locales/translations/ua.json [hunk 5]
```diff
@@ -39,5 +39,53 @@
...
+    "country": {
+      "Label": "Країна",
```
Comment (horba) [inline front/src/locales/translations/ua.json]: it is way to promote copypast in future. Is is vocabulary. Keys can be logical, but not ui-based. No label, options key

File: front/src/views/user/user-settings/user-settings.view.js [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+          firstName: this.firstName,
+          email: this.email,
+          gender: this.gender,
+          language: this.language,
+          country: this.country
+        })
+        .then(() => console.log('Settings updated'))
+        .catch(error => {
+          this.emailAlreadyTaken = error.response.data === 'Email already taken';
+        });
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: action in store

File: front/src/views/user/user-settings/user-settings.view.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  components: {
+    MmtTextInput,
+    MmtSelect,
+    MmtSubmitButton
+  },
+  data () {
+    return {
+      genders: [],
+      languages: [],
+      countries: [],
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: are you sure that it is not in store already?

File: front/src/views/user/user-settings/user-settings.view.js [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+        })
+        .then(() => console.log('Settings updated'))
+        .catch(error => {
+          this.emailAlreadyTaken = error.response.data === 'Email already taken';
+        });
+    },
+    fillLoadedFields (data) {
+      this.genders = this.convertToSelectCompatible(data.genders, 'gender');
+      this.languages = this.convertToSelectCompatible(data.languages, 'language');
+      this.countries = this.convertToSelectCompatible(data.countries, 'country');
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: it is vocabularies that are used not only in your component, and each time front will request the same data

Comment (horba) [reply front/src/views/user/user-settings/user-settings.view.js]: also I can't understand why you need converters

File: front/src/views/user/user-settings/user-settings.view.js [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+      this.lastName = data.lastName;
+    },
+    convertToSelectCompatible (arrOfValues, fieldType) {
+      return arrOfValues.map(id => (
+        {
+          value: id,
+          text: this.$t(`userSettings.${fieldType}.options.${id}`)
+        }));
+    }
+  },
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: flexibility vs maintainable

File: front/src/views/user/user-settings/user-settings.view.js [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+  created () {
+    axios.get(`${serverPath}api/userSettings`)
+      .then(({ data }) => {
+        this.fillLoadedFields(data);
+        this.isLoaded = true;
+      });
+  }
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: action in store

File: front/src/views/user/user-settings/user-settings.view.js [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    MmtSelect,
+    MmtSubmitButton
+  },
+  data () {
+    return {
+      genders: [],
+      languages: [],
+      countries: [],
+      lastName: '',
+      firstName: '',
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: don't we need these data anywhere in application?

File: front/src/views/user/user-settings/user-settings.view.js [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+      email: '',
+      gender: '',
+      language: '',
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: is these one connected whith localization? Should be

File: front/src/views/user/user-settings/user-settings.view.js [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+    MmtSubmitButton
+  },
+  data () {
+    return {
+      genders: [],
+      languages: [],
+      countries: [],
+      lastName: '',
+      firstName: '',
+      email: '',
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: actually, email data is usually necessary not only for user settings

File: server/WebAPI/DTO/UserSettings/UserSettingsResponse.cs [hunk 14]
```diff
@@ -0,0 +1,35 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Entities.Models;
+using Entities.Models.Enums;
+
+namespace WebAPI.DTO.UserSettings
+{
+    public class UserSettingsResponse
```
Comment (horba) [inline server/WebAPI/DTO/UserSettings/UserSettingsResponse.cs]: UserSettingsDTO, see my comment for Irina

**Review by horba — CHANGES_REQUESTED** [review]
component - is about user interaction, logic part and data management are in store

File: front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html (hunk 2 above)
Comment (demosglok) [reply front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html]: why are we binding to attrs here?

File: front/src/components/controls/mmt-select/mmt-select.component.js [hunk 15]
```diff
@@ -0,0 +1,9 @@
+export default {
+  inheritAttrs: false,
+  computed: {
+    attrs () {
```
Comment (demosglok) [inline front/src/components/controls/mmt-select/mmt-select.component.js]: do we really need it?

File: front/src/components/controls/mmt-submit-button/mmt-submit-button.component.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+export default {
+  computed: {
+    attrs () {
+      const { rounded, depressed, large, ...other } = this.$attrs;
```
Comment (demosglok) [inline front/src/components/controls/mmt-submit-button/mmt-submit-button.component.js]: why do we need it?

File: front/src/components/controls/mmt-text-input/mmt-text-input.component.css [hunk 17]
```diff
@@ -1,8 +1,16 @@
 * >>> * {
-  font-size: 12px;
   font-weight: normal;
 }
 
+label,
+.mmt-text-input >>> .v-messages__message {
+  font-size: 12px;
```
Comment (demosglok) [inline front/src/components/controls/mmt-text-input/mmt-text-input.component.css]: will css-variable be more preferable here?

File: front/src/views/user/user-settings/user-settings.view.html [hunk 18]
```diff
@@ -0,0 +1,22 @@
+<v-card class="user-settings-card" flat tile min-height="100%" color="primary lighten-3" :disabled="!isLoaded" >
+  <v-card-title>
+    <h5 class="form-header">{{$t('userSettings.title')}}</h5>
+  </v-card-title>
+  <v-card-text>
+    <v-form class="settings-form" ref="form" v-model="formValidity">
```
Comment (demosglok) [inline front/src/views/user/user-settings/user-settings.view.html]: do we really need "ref" here?

File: server/WebAPI/Controllers/UserSettingsController.cs [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+        {
+            _userRepository = userRepository;
+        }
+
+        [HttpGet]
+        public IActionResult GetSettings()
+        {
+            if (!Int32.TryParse(HttpContext.User.Claims.FirstOrDefault(x => x.Type == ClaimTypes.Name)?.Value, out var id))
+            {
+                return Forbid();
```
Comment (demosglok) [inline server/WebAPI/Controllers/UserSettingsController.cs]: Forbidden?

File: server/WebAPI/Controllers/UserSettingsController.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+            if (user == null)
+            {
+                return StatusCode(500);
+            }
+
+            var userWithThisEmail = _userRepository.FindUserByEmail(userSettings.Email);
+
+            if (userWithThisEmail != null && userWithThisEmail.Id != id)
+            {
+                return BadRequest("Email already taken");
```
Comment (demosglok) [inline server/WebAPI/Controllers/UserSettingsController.cs]: Email IS taken:)

**Review by demosglok — CHANGES_REQUESTED** [review]

File: front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html (hunk 2 above)
Comment (horba) [reply front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html]: концепция прозрачного компонента враппера

Comment (horba) [reply front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html]: правда, здесь я не вижу в ней смысла

File: server/WebAPI/DTO/UserSettings/UserSettingsResponse.cs (hunk 14 above)
Comment (horba) [reply server/WebAPI/DTO/UserSettings/UserSettingsResponse.cs]: http://joxi.ru/Vm6ok6Vsj9eYX2

File: front/src/config/config.dev.json (hunk 3 above)
Comment (valerachuk) [reply front/src/config/config.dev.json]: Should I create a config file for validation rules ?

File: front/src/views/user/user-settings/user-settings.view.js (hunk 12 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js]: These fields are containing ids, translation generates here: convertToSelectCompatible()

File: front/src/config/config.dev.json (hunk 3 above)
Comment (horba) [reply front/src/config/config.dev.json]: it is not config. if you want you can create file with constants

File: front/src/views/user/user-settings/user-settings.view.js (hunk 12 above)
Comment (horba) [reply front/src/views/user/user-settings/user-settings.view.js]: yes, as I said it may make feeling of flexibility, but from other side it makes code more difficult for reading and support

File: front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html (hunk 2 above)
Comment (valerachuk) [reply front/src/components/controls/mmt-submit-button/mmt-submit-button.component.html]: Не нужно будет по несколько раз прописывать одни и те же стили и атрибуты форматирования. Такая кнопка есть в форме логина и в форме настроек. Тогда я верну ее обратно

File: server/WebAPI/Controllers/UserSettingsController.cs (hunk 19 above)
Comment (valerachuk) [reply server/WebAPI/Controllers/UserSettingsController.cs]: Yes, user can't get settings if id is not represented in JWT

Comment (horba) [reply server/WebAPI/Controllers/UserSettingsController.cs]: Дмитрий говорит, что называться должно Forbidden, а не Forbid

Comment (valerachuk) [reply server/WebAPI/Controllers/UserSettingsController.cs]: В ControllerBase нет метода Forbidden, есть Forbid

**Comment by valerachuk** [comment]
image
Some warnings :(

File: front/src/main.js [hunk 21]
```diff
@@ -6,6 +6,7 @@ import vuetify from './plugins/vuetify';
 import i18n from './plugins/i18n';
 import axios from 'axios';
 import { Default, Authorize } from '@layout';
+import { fetchServer } from '@/services/fetchServer.js';
```
Comment (horba) [inline front/src/main.js]: add path `@services`

File: front/src/main.js [hunk 22]
```diff
@@ -35,6 +36,11 @@ new Vue({
       }
       return Promise.reject(error);
     });
+
+    if (this.$store.getters.isLoggedIn) {
+      // get userSettings data from server
```
Comment (horba) [inline front/src/main.js]: these comments usually means that your code is understandable

Comment (horba) [reply front/src/main.js]: 1) it is really strange name fetchServer if you fetch actually user settings
2) i don't agree that only one method is enough and has sense to place into separate service

File: front/src/services/fetchServer.js [hunk 23]
```diff
@@ -0,0 +1,5 @@
+import store from '@store';
+
+export const fetchServer = () => {
+  store.dispatch('userSettings/loadUserSettings');
```
Comment (horba) [inline front/src/services/fetchServer.js]: why did you do it? it is really confused me

File: front/src/views/sign-in/sign-in.view.js [hunk 24]
```diff
@@ -21,6 +23,7 @@ export default {
         password: this.password
       })
         .then(() => {
+          fetchServer();
           this.$router.push('/');
```
Comment (horba) [inline front/src/views/sign-in/sign-in.view.js]: routing push better to place into component. it is interaction.

File: front/src/views/user/user-settings/user-settings.view.js [hunk 25]
```diff
@@ -0,0 +1,149 @@
...
+        },
+        {
+          text: this.$t('locale.ua'),
+          value: 3
+        }
+      ];
+    },
+    ...mapGetters('userSettings', [ 'userSettings' ])
+  },
+  watch: {
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js:106]: why did you need watch?

File: front/src/views/user/user-settings/user-settings.view.js [hunk 26]
```diff
@@ -0,0 +1,149 @@
...
+    userSettings (data) {
+      if (!this.isLoaded) {
+        this.fillLoadedFields(data);
+      }
+    }
+  },
+  methods: {
+    ...mapActions('userSettings', [ 'updateUserSettings' ]),
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: why need you map store methods?

File: front/src/views/user/user-settings/user-settings.view.js [hunk 27]
```diff
@@ -0,0 +1,149 @@
...
+          lastName: this.lastName,
+          firstName: this.firstName,
+          email: this.email,
+          gender: this.gender,
+          languageId: this.languageId,
+          countryId: this.countryId
+        }
+      )
+        .catch(error => {
+          this.emailIsAlreadyTaken = error.response.data === 'Email is already taken';
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: it should be checked not via text. you can use error numbers or others

File: front/src/views/user/user-settings/user-settings.view.js [hunk 28]
```diff
@@ -0,0 +1,149 @@
...
+        });
+    },
+    fillLoadedFields (data) {
+      this.gender = data.gender;
+      this.languageId = data.languageId;
+      this.countryId = data.countryId;
+      this.email = data.email;
+      this.firstName = data.firstName;
+      this.lastName = data.lastName;
+      this.isLoaded = true;
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]: why did you need isLoaded?

File: front/src/views/user/user-settings/user-settings.view.js [hunk 29]
```diff
@@ -0,0 +1,149 @@
...
+    }
+  },
+  created () {
+    // if settings are available in store - fill fields
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js:122]: you have computed properties from store.

**Review by horba — CHANGES_REQUESTED** [review]
Might, you need to read about store and computed properties

File: front/src/views/user/user-settings/user-settings.view.js (hunk 28 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js]: To disable form, while form data is loading

File: front/src/views/user/user-settings/user-settings.view.js [hunk 30]
```diff
@@ -0,0 +1,149 @@
...
+    if (this.userSettings != null) {
+      this.fillLoadedFields(this.userSettings);
```
Comment (horba) [inline front/src/views/user/user-settings/user-settings.view.js]:
```
data() {
editedUserSettings: {}
}
```
=======
`this.editedUserSettings = Object.assign({}, this.userSettings)`

=============

`<form v-if="editedUserSettings"/>`

https://developer.mozilla.org/ru/docs/Web/JavaScript/Reference/Global_Objects/Object/assign

File: front/src/store/modules/userSettings.js [hunk 31]
```diff
@@ -0,0 +1,31 @@
...
+        });
+    },
+    updateUserSettings ({ commit }, formData) {
+      return axios.put(`${serverPath}api/userSettings`, formData)
+        .then(() => commit('SET_USER_SETTINGS', formData));
+    }
+  },
+  getters: {
+    userSettings (state) {
+      return state.userSettings;
```
Comment (horba) [inline front/src/store/modules/userSettings.js]: do we really need this getter?

File: front/src/views/sign-in/sign-in.view.js [hunk 32]
```diff
@@ -21,6 +23,7 @@ export default {
         password: this.password
       })
         .then(() => {
+          fetchServer();
```
Comment (horba) [inline front/src/views/sign-in/sign-in.view.js]: why here?
1) you can just call dispatch action and it will be more clear like store.dispatch('userSettings/loadUserSettings'); 
2) why here we need to init user settings?it is last place where i will looking for init of user data. i expect that we call it in store on successfull authorize or from base authorized layout, but not here

File: front/src/views/user/user-settings/user-settings.view.js (hunk 25 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js:106]: To fill form fields when data becomes available

File: front/src/views/user/user-settings/user-settings.view.js (hunk 26 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js:117]: To make code more readable

File: front/src/services/fetchServer.js (hunk 23 above)
Comment (valerachuk) [reply front/src/services/fetchServer.js]: In future, here we can add actions with the same functionality

File: front/src/store/modules/userSettings.js (hunk 31 above)
Comment (valerachuk) [reply front/src/store/modules/userSettings.js]: No, I forget to delete it

File: front/src/views/sign-in/sign-in.view.js (hunk 32 above)
Comment (valerachuk) [reply front/src/views/sign-in/sign-in.view.js]: 1. already here
2. ok, I will place it inside layout

File: front/src/views/user/user-settings/user-settings.view.js (hunk 30 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js]: > this.editedUserSettings = Object.assign({}, this.userSettings)

If form creates before its data loads, editedUserSettings will be empty. So its better to place this line inside watch

File: front/src/store/modules/userSettings.js (hunk 31 above)
Comment (valerachuk) [reply front/src/store/modules/userSettings.js:28]: Hm, but watch needs it...

File: front/src/views/user/user-settings/user-settings.view.js (hunk 30 above)
Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js:145]: > !!editedUserSettings
is always true

Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js]: Object is always true...
I can use isLoaded or editedUserSettings.email to check form loading status

Comment (valerachuk) [reply front/src/views/user/user-settings/user-settings.view.js]: Used computed property: 

```javascript
isLoaded () {
      return !!Object.keys(this.editedUserSettings).length;
    }
```

**Review by horba — APPROVED** [review]
